src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -218,26 +219,30 @@ class ImageReplayerAdminSocketHook : public AdminSocketHook {
     for (Commands::const_iterator i = commands.begin(); i != commands.end();
 	 ++i) {
       (void)admin_socket->unregister_command(i->first);
-      delete i->second;
     }
     commands.clear();
   }
 
   bool call(std::string command, cmdmap_t& cmdmap, std::string format,
 	    bufferlist& out) override {
-    Mutex::Locker locker(lock);
-    Commands::const_iterator i = commands.find(command);
-    assert(i != commands.end());
+    CommandRef command_ref;
+    {
+      Mutex::Locker locker(lock);
+      Commands::const_iterator i = commands.find(command);
+      assert(i != commands.end());
+      command_ref = i->second;
+    }
     Formatter *f = Formatter::create(format);
     stringstream ss;
-    bool r = i->second->call(f, &ss);
+    bool r = command_ref->call(f, &ss);
```

Comment:
Fair enough -- I am just concerned about the edge condition where ```ImageReplayer``` is shutting down. It looks like it will now be possible for a call to start, grab a shared pointer, release the lock, <destruct image replayer>, invoke callback on destructed image replayer.